library/kotlin/src/io/envoyproxy/envoymobile/GRPCResponseHandler.kt

```diff
@@ -0,0 +1,196 @@
+
+      processState = processData(byteBufferedOutputStream, processState, closure)
+    }
+
+    return this
+  }
+
+  /**
+   * Called when response metadata is received by the stream.
+   *
+   * @param metadata the metadata of a response.
+   * @param endStream true if the stream is complete.
+   * @return ResponseHandler, this ResponseHandler.
+   */
+  fun onMetadata(closure: (metadata: Map<String, List<String>>) -> Unit): GRPCResponseHandler {
+    underlyingHandler.onMetadata(closure)
+    return this
+  }
+
+  /**
+   * Specify a callback for when trailers are received by the stream.
+   * If the closure is called, the stream is complete.
+   *
+   * @param closure: Closure which will receive the trailers.
+   * @return ResponseHandler, this ResponseHandler.
+   */
+  fun onTrailers(closure: (trailers: Map<String, List<String>>) -> Unit): GRPCResponseHandler {
+    underlyingHandler.onTrailers(closure)
+    return this
+  }
+
+  /**
+   * Specify a callback for when an internal Envoy exception occurs with the stream.
+   * If the closure is called, the stream is complete.
+   *
+   * @param closure: Closure which will be called when an error occurs.
+   * @return ResponseHandler, this ResponseHandler.
+   */
+  fun onError(closure: (error: EnvoyError) -> Unit): GRPCResponseHandler {
+    this.errorClosure = closure
+    underlyingHandler.onError(closure)
+    return this
+  }
+
+  /**
+   * Specify a callback for when the stream is canceled.
+   * If the closure is called, the stream is complete.
+   *
+   * @param closure: Closure which will be called when the stream is canceled.
+   * @return ResponseHandler, this ResponseHandler.
+   */
+  fun onCanceled(closure: () -> Unit): GRPCResponseHandler {
+    underlyingHandler.onCanceled(closure)
+    return this
+  }
+
+  /**
+   * Recursively processes a buffer of data, buffering it into messages based on state.
+   * When a message has been fully buffered, `onMessage` will be called with the message.
+   *
+   * @param bufferedStream The buffer of data from which to determine state and messages.
+   * @param processState The current process state of the buffering.
+   * @param onMessage Closure to call when a new message is available.
+   */
+  private fun processData(
+      bufferedStream: ByteArrayOutputStream,
+      processState: ProcessState,
+      onMessage: (byteBuffer: ByteBuffer) -> Unit): ProcessState {
+
+    var nextState = processState
+
+    when (processState) {
+      is ProcessState.CompressionFlag -> {
+        val byteArray = bufferedStream.toByteArray()
+        if (byteArray.size < MESSAGE_HEADING_OFFSET) {
+          // We don't have enough information so we'll just return
+          return ProcessState.CompressionFlag
+        }
+
+        val compressionFlag = byteArray[0]
+        // TODO: Support gRPC compression https://github.com/lyft/envoy-mobile/issues/501
+        if (compressionFlag.compareTo(0) != 0) {
+          errorClosure(
```

Comment:
The important detail here is that we still keep the stream alive 😬 

If we can safely close the stream, that would be much much better. 

I just left it like this for now since I do consider this a volatile feature and the behavior is a bit up in the air with respect to messages which are gRPC compressed. If we appropriately close the stream, it feels more of a intentional design to exclude compression

Reply:
What if we just call `close()` here and call the consumer with `onError`? I think that'd be a good way to shut down safely

Reply:
We also don't have a `close()` handler in the response case

Reply:
aha right

Reply:
I just realized why I didn't do this originally on iOS. Storing the `errorClosure` on the `GRPCResponseHandler` is problematic because nothing necessarily keeps the `GRPCResponseHandler` in memory since only the closures need to stick around. When I ended up using the `GRPCResponseHandler` in practice, this was totally fine because the closures passed to the underlying callbacks have their own lifecycle. Maybe we should just assert here instead?

Reply:
https://github.com/lyft/envoy-mobile/pull/515